packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -69,145 +69,248 @@ import { applyEventListeners } from './modules/events';
-// Custom Element that is inserted via a template.
-export const ElementHook: Hooks<VElement> = {
-    create: (vnode) => {
-        const {
-            sel,
-            owner,
-            data: { svg },
-        } = vnode;
+export function patchChildren(c1: VNodes, c2: VNodes, parent: ParentNode): void {
+    if (hasDynamicChildren(c2)) {
+        updateDynamicChildren(c1, c2, parent);
+    } else {
+        updateStaticChildren(c1, c2, parent);
+    }
+}
 
-        const namespace = isTrue(svg) ? SVG_NAMESPACE : undefined;
-        const elm = createElement(sel, namespace);
-
-        linkNodeToShadow(elm, owner);
-        fallbackElmHook(elm, vnode);
-        vnode.elm = elm;
-
-        patchElementPropsAndAttrs(null, vnode);
-    },
-    update: (oldVnode, vnode) => {
-        patchElementPropsAndAttrs(oldVnode, vnode);
-        patchChildren(vnode.elm!, oldVnode.children, vnode.children);
-    },
-    insert: (vnode, parentNode, referenceNode) => {
-        insertNode(vnode, parentNode, referenceNode);
-        createChildrenHook(vnode);
-    },
-    move: insertNode,
-    remove: (vnode, parentNode) => {
-        removeNode(vnode, parentNode);
-        removeChildren(vnode);
-    },
-};
-
-export const CustomElementHook: Hooks<VCustomElement> = {
-    create: (vnode) => {
-        const { sel, owner } = vnode;
-        const UpgradableConstructor = getUpgradableConstructor(sel);
-        /**
-         * Note: if the upgradable constructor does not expect, or throw when we new it
-         * with a callback as the first argument, we could implement a more advanced
-         * mechanism that only passes that argument if the constructor is known to be
-         * an upgradable custom element.
-         */
-        let vm: VM | undefined;
-        const elm = new UpgradableConstructor((elm: HTMLElement) => {
-            // the custom element from the registry is expecting an upgrade callback
-            vm = createViewModelHook(elm, vnode);
-        });
-
-        linkNodeToShadow(elm, owner);
-        vnode.elm = elm;
-
-        if (vm) {
-            allocateChildren(vnode, vm);
-        } else if (vnode.ctor !== UpgradableConstructor) {
-            throw new TypeError(`Incorrect Component Constructor`);
-        }
-        patchElementPropsAndAttrs(null, vnode);
-    },
-    update: (oldVnode, vnode) => {
-        patchElementPropsAndAttrs(oldVnode, vnode);
-        const vm = getAssociatedVMIfPresent(vnode.elm);
-        if (vm) {
-            // in fallback mode, the allocation will always set children to
-            // empty and delegate the real allocation to the slot elements
-            allocateChildren(vnode, vm);
+function patch(n1: VNode, n2: VNode, parent: ParentNode) {
+    if (n1 === n2) {
+        return;
+    }
+
+    // FIXME: When does this occurs, is it possible with LWC?
+    if (!isSameVnode(n1, n2)) {
+        unmount(n1, parent, true);
+        mount(n2, parent, nextSibling(n1.elm));
+        return;
+    }
```

Comment:
None of our Jest/Karma tests hit this condition, but I'm not sure if it's impossible to hit...

Reply:
I think it's impossible to hit this condition. Here's why:

## Call sites that already check `isSameVnode()`

Most places that call `patch()` already check for `isSameVnode()`, e.g.:

https://github.com/salesforce/lwc/blob/0094a68292cb55af2c2dc519c194e7681b63599b/packages/%40lwc/engine-core/src/framework/rendering.ts#L604-L605

## Call site that verifies `key` and `sel`

As for the call sites that _don't_ check, there are two. First, there's this one:

https://github.com/salesforce/lwc/blob/0094a68292cb55af2c2dc519c194e7681b63599b/packages/%40lwc/engine-core/src/framework/rendering.ts#L636

But this one has already guaranteed that `newStartVnode` and `elmToMove` have the same key:

https://github.com/salesforce/lwc/blob/0094a68292cb55af2c2dc519c194e7681b63599b/packages/%40lwc/engine-core/src/framework/rendering.ts#L624

... and the same `sel`:

https://github.com/salesforce/lwc/blob/0094a68292cb55af2c2dc519c194e7681b63599b/packages/%40lwc/engine-core/src/framework/rendering.ts#L632-L635

... which is the definition of `isSameVnode`:

https://github.com/salesforce/lwc/blob/0094a68292cb55af2c2dc519c194e7681b63599b/packages/%40lwc/engine-core/src/framework/vnodes.ts#L89-L91

## Call site that relies on static children ordering

The other call site is in `updateStaticChildren`:

https://github.com/salesforce/lwc/blob/0094a68292cb55af2c2dc519c194e7681b63599b/packages/%40lwc/engine-core/src/framework/rendering.ts#L701

... in which case, we are dealing with two arrays of children that are already guaranteed to be the same (static), and we are taking the same item at the same index for both:

https://github.com/salesforce/lwc/blob/0094a68292cb55af2c2dc519c194e7681b63599b/packages/%40lwc/engine-core/src/framework/rendering.ts#L693-L695

So I think this code is safe to remove!

Reply:
It turns out that the last scenario (`updateStaticChildren`) _can_ hit this code path, if the `key` is defined outside of an iteration: https://github.com/salesforce/lwc/issues/2697.

We can add code to handle this particular scenario.

Reply:
Good catch! I added a bunch of thoughts about this issue on the issue: https://github.com/salesforce/lwc/issues/2697#issuecomment-1042737070